Script Based Tokenizer (#2)
meilisearch/specifications: text/0001-script-based-tokenizer.md

    fn new(inner: &'a str) -> Self { unimplemented!() }
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token<'a>;

    /// return the next token calling the `next` method of the internal tokenizer,
    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
    /// if no iternal tokenizer is chosen, the method return None
    fn next(&mut self) -> Option<Self::Item> { unimplemented!() }
}
```

## Token

```rust
/// script of a token (https://docs.rs/whatlang/0.10.0/whatlang/enum.Script.html)
pub type Script = whatlang::Script;

/// atomic item returned by `Tokenizer::next()`,
/// it determine the type of the token and contains a `WordSlice`
pub enum Token<'a> {
    /// the token is a word,
    /// meaning that it should be indexed as an important part of the document
    Word(WordSlice<'a>),
    /// the token is a stop word,
    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
    StopWord(WordSlice<'a>),
    /// the token is a separator,
    /// meaning that it shouln't be indexed but used to determine word proximity
    Separator(WordSlice<'a>)

> }

Reviewer: > we don't have to bother with a match where we don't care about the TokenType itself, but only the token (&str).

you don't have to, Token exposes a `txt` method.

> I am also not sure that the WordSlice supports the fact that multiple words can live at the same place

What does that mean?

/// The script, the char_index and the content of the token
pub struct WordSlice<'a> {
    /// content of the token
    pub word: &'a str,

>     /// index of the first character of the token in the whole document
>     pub char_index: usize,

Reviewer: ```suggestion
```

- detect the script of the start of the document, tokenize with specialized lexer while script is the same, switch lexer if the script change

#### Tokenizer

```rust
use crate::token::Token;
use crate::internal_tokenizer::InternalTokenizer;

struct Tokenizer<'a> {
    /// script specialized tokenizer, this can be switched during
    /// document tokenization if the document contains several scripts
    current_tokenizer: Option<Box<dyn InternalTokenizer<'a>>>,
    /// current character index in the document
    current_char_index: u64,
    /// reference on the document content
    inner: &'a str,
}

impl<'a> Tokenizer<'a> {
    /// create a new tokenizer detecting script
    /// and chose the specialized internal tokenizer
    fn new(inner: &'a str) -> Self { unimplemented!() }
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token<'a>;

    /// return the next token calling the `next` method of the internal tokenizer,
    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
    /// if no iternal tokenizer is chosen, the method return None

>     fn next(&mut self) -> Option<Self::Item> { unimplemented!() }

Reviewer: @LegendreM @Kerollmops 
The implementation of `next` for the `Tokenizer` is in charge of selecting the appropriate tokenizer to use to the section of text. There are two way to go about this:
- Hardcoding the selection in a `match` => straightforward implementation
- Using a map that links scripts to tokenizer. This would allow for user configurable choice of tokenizer in the future, (wasm, ngram...) for specific scripts.

> }

Reviewer: we will need to use a StreamingIterator to return tokens that reference the Interator itself.

```

#### Token

```rust
/// script of a token (https://docs.rs/whatlang/0.10.0/whatlang/enum.Script.html)
pub type Script = whatlang::Script;

/// atomic item returned by `Tokenizer::next()`
pub struct Token<'a> {
    word: Lexeme<'a>,
    script: Script,

> }

Reviewer: @Kerollmops @LegendreM 
We want to allow for overlapping tokens (i.e tokens that may refer to the same word position). There are multiple solutions to that:
- (prefered) add a `word_position` field to the WordSlice, multiple `WordSlice` can have the same word index.
- Make `Lexme` contain a Cow<[WordSlice]> that contains all the word slices for the same word position, this implies that token's `txt` method returns a collection of `&str` which may not be very ergonomic?
- Make `Token` contain a `Cow<[Lexeme]>`, similar to before, a bit more general, same issue

impl<'a> Token<'a> {
    fn txt(&self) -> &str { unimplemented!() }
}

/// Determine the type of the token and contains a `WordSlice`
pub enum Lexeme<'a> {
    /// the token is a word,
    /// meaning that it should be indexed as an important part of the document
    Word(WordSlice<'a>),
    /// the token is a stop word,
    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
    StopWord(WordSlice<'a>),
    /// the token is a separator,
    /// meaning that it shouldn't be indexed but used to determine word proximity
    Separator(WordSlice<'a>)
}

/// The script, the char_index and the content of the token
pub struct WordSlice<'a> {
    /// content of the token
    pub word: &'a str,
    /// index of the first character of the token in the whole document
    pub char_index: usize,
}
```

> 

Reviewer: @LegendreM @Kerollmops 
## Normalization

Normalization is an issue that hasn't been discussed. Some scripts will require words to be normalized (lowercased, deunicoded) but not all. It seems natural that this task is handled by the tokenizer. Two options:
- The token has a `normalize` method that returns a Cow<str>, and that know how to perform the normalization for this token with the information it has.
- An option `normalize` is given to the `Tokenizer`, and tokens contains a `Cow<str>` instead of a `&str`.

- Title: Script Based Tokenizer
- Start Date: 2020-10-27
- specification PR: meilisearch/specifications#2
- Meilisearch Issue: meilisearch/Meilsearch#624

## Feature Description and Interaction

### Summary

> Tokenizer's principal role is to split documents into words (tokens) so each document can be indexed by it's contained words. It is also needed to split a search query into tokens, and search in the words index.

Reviewer 2: About the first part of the sentence, is this a reference to reverse-indexing?

About the second part of the sentence, I do not understand it :s 
` It is also needed to split a search query into tokens, and search in the words index.`

This is my suggestion on the summary:
```suggestion
Tokenizer's principal role is to split documents into words (tokens). Each word becomes a key that has a list of all the documents in which it is present as its value. This is called [reverse-index](https://semify.com/?Reverse-Index&AID=1431#:~:text=A%20reverse%20index%20is%20a,inserting%20them%20into%20the%20index.).
The same tokenizer is used to split a search query into tokens. Those tokens will in turn be used by MeiliSearch to search for matching documents inside the reverse-index previously created.

Some languages do not split the same way as others, the aim of this tokenizer is to create a more diverse tokenizer. 
```

### Motivation

It turns out that splitting texts into tokens is a complicated task, especially when you need to handle more than a single language. While the task for english may seem trivial at first glance (mostly split white space), Chinese tokenization, for example is completely different (no spaces between words, groups of 1, 2 or 3 characters...).

> Actual Tokenizer is naive and partially handle Latin Scripts only, lot of users ask for a smarter tokenizer to handle their own languages.
> We should implement a new tokenizer which detect script and tokenize based on it.

Reviewer 2: ```suggestion
The Tokenizer we have now is naive and partially handles Latin Scripts only, a lot of users ask for a smarter tokenizer to handle their own languages.
We should implement a new tokenizer which detects script and tokenize based on it.
```

  > in a R&D project, @qdequele was able to detect language based on stop word distribution in a text.
  > If, in a latin script, there is lot of French stop words then the text language is probably french.

**other solution that advertise multilingual support:**
- Sonic uses whatlang to peform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer
	- Sonic uses whatlang to detect the languages, but don't acutally seems to use it to segment the text. It simply uses unicode segmentation, I can't really explain what they actually do with the language information.
- tantivy advertise good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer
	- Tantivy is similar to elastic in the sense that you can setup a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a colleciton of tokenizer to choose from. Token are rather simple and do not contain any metadata, except for it's position.
- How elastic search handle it: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html:

> 	- Algolia proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the collowing components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)

Reviewer 3: ```suggestion
	- Elasticsearch proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the collowing components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)
```

- Algolia:

> 	- Algolia seems to use default word segmentation use user configurable separators. The use can also set the `queryLanguage` or provide a list of `stopWords`, but it is unclear how this work under the hood for languages such as Mandarin. They also word decompounging, meaning that they can split words that are mode of multiple words (mostly germanic/nordic languages apparently).

Reviewer 3: They explain everything on the [doc](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/).

```suggestion
    - Algolia use multiple techniques to handle tokenization. They start by [normalize](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the data (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurials. 
```

      // get the longest normalized token
      let token = token_goup.normalized().max_by(|a, b| a.token_len().cmp(b.token_len()));
      if token.kind() == TokenKind::Word {
        let normalized_token = token.text();
        let to_highlight = words.contains(&normalized_token);
        if to_highlight { value.push_str("<mark>") }
        value.push_str(token.original.text());
        if to_highlight { value.push_str("</mark>") }
      } else {
        value.push_str(token);
      }
    }
  }
}
```

```rust
// original
fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
    for (_key, value) in record.iter_mut() {
        let old_value = mem::take(value);
        for (token_type, token) in simple_tokenizer(&old_value) {
            if token_type == TokenType::Word {
                let lowercase_token = token.to_lowercase();
                let to_highlight = words.contains(&lowercase_token);
                if to_highlight { value.push_str("<mark>") }
                value.push_str(token);
                if to_highlight { value.push_str("</mark>") }
            } else {
                value.push_str(token);
            }
        }
    }
}

> ```

Reviewer 3: Could you explain a little bit more about the difference? This seems to not be possible on the current MeiliSearch. Because at the moment, the engine cannot give us all words (variations) that matched on documents. Is this a change to make? If yes, could you write it?

    }
}
```

```rust
// original
for (pos, token) in simple_tokenizer(&content).filter_map(only_token).enumerate().take(MAX_POSITION) {
    let word = token.to_lowercase();
    let position = (attr as usize * MAX_POSITION + pos) as u32;
    words_positions.entry(word).or_insert_with(SmallVec32::new).push(position);
}
```

### Impact on documentation

This feature should not impact the meilisearch user documentation,
in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.

## Technical Specifications

### Architecture

As the actual tokenizer, this tokenizer will be a standolone library,
and will be used by meilisearch like another one.
We decided to put it in an other repository to be able to use it in meilisearch/Meilisearch and in the future core-engine.

### Implementation Details

We could have 2 approach of tokenization based on the script,
the main difficulty is to handle multi-script documents
which need to segment each part of script and tokenize them differently,
theses 2 approach are:
- first iterate over the whole document, splitting it by script, and after, tokenize each part with specialized lexer

> - detect the script of the start of the document, tokenize with specialized lexer while script is the same, switch lexer if the script change

Reviewer 3: Still not clear for me, what is the difference between these two points?

#### Tokenizer

```rust
use crate::token::Token;
use crate::internal_tokenizer::InternalTokenizer;

struct Tokenizer<'a> {
    /// script specialized tokenizer, this can be switched during
    /// document tokenization if the document contains several scripts
    current_tokenizer: Option<Box<dyn InternalTokenizer<'a>>>,
    /// current character index in the document
    current_char_index: u64,
    /// reference on the document content
    inner: &'a str,
}

impl<'a> Tokenizer<'a> {
    /// create a new tokenizer detecting script
    /// and chose the specialized internal tokenizer
    fn new(inner: &'a str) -> Self { unimplemented!() }

> }

Reviewer 3: How do we configure stop words?

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token<'a>;

    /// return the next token calling the `next` method of the internal tokenizer,
    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
    /// if no iternal tokenizer is chosen, the method return None

>     fn next(&mut self) -> Option<Self::Item> { unimplemented!() }

Reviewer 3: Are you talking about switching the tokenizer inside the same attribute? It seems too much. I think we can put the limit at one script per field.

- Algolia:
    - Algolia use multiple techniques to handle tokenization. They start by [normalizing](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the text (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurals. 

### Explanation

> In order to use the tokenizer, all the user has to do is to instantiate a `Tokenizer`, call `tokenize(&str)` on it and iterate over the emitted tokens:

Reviewer 2: Could you add a mention to `whatlang` as it is not clear that we use that technique for tokenization.

Reviewer: we don't really

```rust
let stop_words = Set::from_iter(["the", "of"].iter()).unwrap();
let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));

let orig = "The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!";
let analyzed = analyzer.analyze(orig);
let mut analyzed = analyzed.tokens();
assert_eq!("the", analyzed.next().unwrap().text());
assert_eq!(" ", analyzed.next().unwrap().text());
assert_eq!("quick", analyzed.next().unwrap().text());
assert_eq!(" ", analyzed.next().unwrap().text());
assert_eq!("(", analyzed.next().unwrap().text());
assert_eq!("\"", analyzed.next().unwrap().text());
assert_eq!("brown", analyzed.next().unwrap().text());
```

The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep it's configuration state and allocations.

> Bellow are example of the integration of the new tokenizer in existing code:

Reviewer 2: ```suggestion
Below are examples of the integration of the new tokenizer in existing code:
```

- Highlight in @kerollmops milli:
```rust
// new tokenizer
fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {

>     let stop_words = Set::from_iter([""].iter()).unwrap();

Author: ```suggestion
   let stop_words = Set::default();
```

        let analyzed = analyzer.analyze(&old_value);
        
        for (original, token) in analyzed.reconstruct() {
            if token.is_word() {
                let to_highlight = words.contains(&token.text());
                if to_highlight { value.push_str("<mark>") }
                value.push_str(original);
                if to_highlight { value.push_str("</mark>") }
            } else {
                value.push_str(original);
            }
        }
    }
}
```
```rust
// original
fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
    for (_key, value) in record.iter_mut() {
        let old_value = mem::take(value);
        for (token_type, token) in simple_tokenizer(&old_value) {
            if token_type == TokenType::Word {
                let lowercase_token = token.to_lowercase();
                let to_highlight = words.contains(&lowercase_token);
                if to_highlight { value.push_str("<mark>") }
                value.push_str(token);
                if to_highlight { value.push_str("</mark>") }
            } else {
                value.push_str(token);
            }
        }
    }
}
```

> As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make it's API close to the previous one.

Reviewer 2: ```suggestion
As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make its API close to the previous one.
```

### Impact on documentation

> This feature should not impact the meilisearch user documentation,
> in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.

Reviewer 2: ```suggestion
This feature should not impact meilisearch users' documentation. 
In future versions, we will probably provide a way to configure tokenizer and this will be discussed in a new specification.
```

## Technical Specifications

### Architecture

The new version of the tokenizer will replace the current version as a standalone library.

### Implementation Details

> We want to support different tokenizers based on the language of the text to be indexed. For this we may need to change the tokenizer we are using while indexing, depending on the language and the script. The Analyzer provides an interface that abstract this need away from the consumer of the tokens.

Reviewer 2: ```suggestion
We want to support different tokenizers based on the language of the text that needs to be indexed. For this, we may need to change the tokenizer we are using while indexing, depending on the language and the script. The Analyzer provides an interface that abstracts this need away from the consumer of the tokens.
```
I think I understand that the Analyzer infers the scripts, but who is `the consumer of the tokens`?

#### Pipeline

> A pipeline contains all the steps needed to tokenize a specific language or script,
> one or several preprocessors to prepare text before tokenization e.g. traditional chinese translation
> or erasing some characters;
> a tokeninizer to transform the text in an iterator of tokens;
> one or several normalizers to normalize tokens for indexation e.g. deunicode or lowercase.

Reviewer 2: ```suggestion
A pipeline is a structure that contains all the steps needed to tokenize a specific language or script:
- One or several preprocessors to prepare text before tokenization e.g. traditional Chinese translation or erasing some characters;
- A tokenizer to transform the text into an iterator of tokens;
- One or several normalizers to normalize tokens for indexation e.g. deunicode or lowercase.
```

```rust
pub struct Pipeline {
    pre_processor: Box<dyn PreProcessor + 'static>,
    tokenizer: Box<dyn Tokenizer + 'static>,
    normalizer: Box<dyn Normalizer + 'static>,
}
```

#### Configuration

> The tokenizer is configured at initialization. A `Default` implementation is provided with the settings we believe suits most cases.

Reviewer 2: ```suggestion
The tokenizer is configured at initialization. A `default` implementation is provided with the settings we believe suits most cases.
```

Reviewer: default is the name of a trait, it start with an uppercase :)

The configuration options, for now, are:
- the `stop_words`

> - the `pipeline_map` that links a `(Script, Language)` to a `Pipeline`. This defaults to the mapping we find best fitting for most cases. This settings should be left as is now, but will allow user defined Pipelines in the future.

Reviewer 2: ```suggestion
- the `pipeline_map` that links a `(Script, Language)` to a `Pipeline`. This defaults to the mapping we find the best fitting for most cases. These settings should be left as is now, but will allow the user to define Pipelines in the future.
```

    ///
    /// ```rust
    /// use meilisearch_tokenizer::{Analyzer, AnalyzerConfig};
    /// use fst::Set;
    /// // defaults to unicode segmenter with identity preprocessor and normalizer.
    /// let stop_words = Set::from_iter([""].iter()).unwrap();
    /// let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
    /// let analyzed = analyzer.analyze("The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!");
    /// let mut tokens = analyzed.tokens();
    /// assert!("the" == tokens.next().unwrap().text());
    /// ```
    pub fn analyze<'t>(&'t self, text: &'t str) -> AnalyzedText<'t, A> { unimplemented!() }
}

/// result of analyzer.analyze() function, made to choose how tokens will be returned.
pub struct AnalyzedText<'a, A> {}

impl<'a, A> AnalyzedText<'a, A>
where
    A: AsRef<[u8]>
{
    /// Returns a `TokenStream` for the Analyzed text.
    pub fn tokens(&'a self) -> TokenStream<'a> { unimplemented!() }

    /// Attaches each token to its corresponding portion of the original text.
    pub fn reconstruct(&'a self) -> impl Iterator<Item = (&'a str, Token<'a>)>  { unimplemented!() }
}
```

> #### Token

Reviewer 2: Could you add a small explanation on the token? While I understand all the previous concepts this one I'm not sure. Is a token a portion of a string?

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SeparatorKind {
    Hard,
    Soft,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TokenKind {
    Word,
    /// the token is a stop word,
    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
    StopWord,
    /// the token is a separator,
    /// meaning that it shouldn't be indexed but used to determine word proximity
    Separator(SeparatorKind),
    Unknown,
}
```

> The emitted token position are relative to the original text, this information is reliable. The `word` is normalized if need be.

Reviewer 2: ```suggestion
The emitted token position is relative to the original text, this information is reliable. The `word` is normalized if need be.
```

```rust
#[derive(Debug, Clone, Default)]
pub struct Token<'a> {
    pub kind: TokenKind,
    pub word: Cow<'a, str>,
    /// index of the first character of the word
    pub char_index: usize,
    /// indexes of start and end of the byte slice
    pub byte_start: usize,
    pub byte_end: usize,
}

impl<'a> Token<'a> {
    // return the normalized version of the token
    pub fn text(&self) -> &str { unimplemented!() }

    // return the size in byte of the original text
    pub fn byte_len(&self) -> usize { unimplemented!() }

    // return the TokenKind of the token
    pub fn kind(&self) -> TokenKind { unimplemented!() }

    // return true if the TokenKind of the token is TokenKind::Word
    pub fn is_word(&self) -> bool { unimplemented!() }

    // return Some(SeparatorKind) if the TokenKind of the token is TokenKind::Separator(SeparatorKind),
    // None if the token is not a separator
    pub fn is_separator(&self) -> Option<SeparatorKind> { unimplemented!() }

    // return true if the TokenKind of the token is TokenKind::StopWord
    pub fn is_stopword(&self) -> bool { unimplemented!() }
}
```

#### Internal Tokenizer trait

> The `InternalTokenizer` traits provides a common interface to adapt other tokenizers to the tokenizer. This allows extensibility of the current tokenizer to other languages.

Reviewer 2: ```suggestion
The `InternalTokenizer` traits provide a common interface to adapt other tokenizers to the tokenizer. This allows the extensibility of the current tokenizer to other languages.
```

> The `InternalTokenizer` traits provides a common interface to adapt other tokenizers to the tokenizer. This allows extensibility of the current tokenizer to other languages.

Reviewer: ```suggestion
The `InternalTokenizer` trait provides a common interface to adapt other tokenizers to the tokenizer. This allows the extensibility of the current tokenizer to other languages.
```

```rust
/// iterator over tokens processed by the specialized tokenizer
pub struct TokenStream<'a> {
    pub(crate) inner: Box<dyn Iterator<Item = Token<'a>> + 'a>
}

impl<'a> Iterator for TokenStream<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next()
    }
}

/// trait defining an internal tokenizer,
/// an internal tokenizer should be a script specialized tokenizer,
/// this should be implemented as an `Iterator` with `Token` as `Item`,
pub trait Tokenizer: Sync + Send {
    /// create the tokenizer based on the given `text` and `char_index`
    fn tokenize<'a>(&self, s: &'a ProcessedText<'a>) -> TokenStream<'a>;
}
```

> - use of jieba for chinese tokenization
> - use of unicode-segmenter for other segmentations

Reviewer 2: ```suggestion
- use of `jieba` for Chinese tokenization
- use of `unicode-segmenter` for other segmentations
```

### Corner Cases

> In some languages like Chinese, there can be multiple "words" extracted that are considered to be at the same position in the text. For example 计算所 gives 计算 and 计算所, that are at the same position in the text but don't have the same length, the tokenizer should support that behavior.

Reviewer 2: ```suggestion
In some languages like Chinese, there can be multiple "words" extracted that are considered to be at the same position in the text. For example, 计算所 gives 计算 and 计算所, that are at the same position in the text but don't have the same length, the tokenizer should support that behavior. See #5.
```

> return several token as the same word_position seems to be over-engineering, we may want to have only 1 token by iteration making API simplier
## Future possibilities

> - We should add a way to configure tokenizer forcing a specific language/script

Reviewer 2: ```suggestion
- We should add a way to configure the tokenizer to enforce a specific language/script
```

- Title: Script Based Tokenizer
- Start Date: 2020-10-27
- specification PR: meilisearch/specifications#2
- Meilisearch Issue: meilisearch/Meilsearch#624

## Feature Description and Interaction

### Summary

> The first step of document indexing in the Meilisearch engine is tokenization. Tokenization in the action of taking a sentence and splitting in units of language called tokens. The tokenization task is highly language dependant and is a critical factor in the quality of the search results.

Reviewer 4: ```suggestion
The first step of document indexing in the Meilisearch engine is tokenization. Tokenization is the action of taking a sentence and splitting it in units of language called tokens. The tokenization task is highly language dependant and is a critical factor in the quality of the search results.
```

### Prior Art and R&D

**tokenization:**
- > **[unicode-segmentation](https://github.com/unicode-rs/unicode-segmentation):**
  > tokenizer which follow the [Standard Annex #29: Unicode Text Segmentation](http://www.unicode.org/reports/tr29/),
  > this tokenizer seems promising for Latin scripts.
- > **[Jieba](https://github.com/messense/jieba-rs):**
  > tokenizer specialized in Chinese languages
- > **[Lindera](https://github.com/lindera-morphology/lindera):**
  > Japanese and Korean

**lang/script detection:**
- > **[whatlang](https://github.com/greyblake/whatlang-rs):**
  > whatlang is able to detect script or/and language in a text,
  > language detection is low but script is acceptable.
  > note: Sonic also uses whatlang to peform the tokenization, it could be interesting to checkout how they do it.
- > **toku (@qdequele):**
  > in a R&D project, @qdequele was able to detect language based on stop word distribution in a text.
  > If, in a latin script, there is lot of French stop words then the text language is probably french.

**other solution that advertise multilingual support:**

> - Sonic uses whatlang to peform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer

Reviewer 4: ```suggestion
- Sonic uses whatlang to perform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer
```

- tantivy advertise good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer
	- Tantivy is similar to elastic in the sense that you can setup a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a colleciton of tokenizer to choose from. Token are rather simple and do not contain any metadata, except for it's position.
- How elastic search handle it: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html:
	- Elastic proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the following components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)
- Algolia:

>     - Algolia use multiple techniques to handle tokenization. They start by [normalize](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the data (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurials. 

Reviewer 4: ```suggestion
    - Algolia use multiple techniques to handle tokenization. They start by [normalizing](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the text (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurals. 
```

in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.

## Technical Specifications

### Architecture

The new version of the tokenizer will replace the current version as a standalone library.

### Implementation Details

We want to support different tokenizers based on the language of the text to be indexed. For this we may need to change the tokenizer we are using while indexing, depending on the language and the script. The Analyzer provides an interface that abstract this need away from the consumer of the tokens.

#### Pipeline

A pipeline contains all the steps needed to tokenize a specific language or script,
one or several preprocessors to prepare text before tokenization e.g. traditional chinese translation
or erasing some characters;
a tokeninizer to transform the text in an iterator of tokens;
one or several normalizers to normalize tokens for indexation e.g. deunicode or lowercase.
```rust
pub struct Pipeline {
    pre_processor: Box<dyn PreProcessor + 'static>,
    tokenizer: Box<dyn Tokenizer + 'static>,
    normalizer: Box<dyn Normalizer + 'static>,
}
```

#### Configuration

> The tokenizer in configured at initialization. A `Default` implementation is provided with the settings we believe suits most cases.

Reviewer 4: ```suggestion
The tokenizer is configured at initialization. A `Default` implementation is provided with the settings we believe suits most cases.
```

    /// let mut tokens = analyzed.tokens();
    /// assert!("the" == tokens.next().unwrap().text());
    /// ```
    pub fn analyze<'t>(&'t self, text: &'t str) -> AnalyzedText<'t, A> { unimplemented!() }
}

/// result of analyzer.analyze() function, made to choose how tokens will be returned.
pub struct AnalyzedText<'a, A> {}

impl<'a, A> AnalyzedText<'a, A>
where
    A: AsRef<[u8]>
{
    /// Returns a `TokenStream` for the Analyzed text.
    pub fn tokens(&'a self) -> TokenStream<'a> { unimplemented!() }

    /// Attaches each token to its corresponding portion of the original text.
    pub fn reconstruct(&'a self) -> impl Iterator<Item = (&'a str, Token<'a>)>  { unimplemented!() }
}
```

#### Token

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SeparatorKind {
    Hard,
    Soft,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TokenKind {
    Word,
    /// the token is a stop word,
    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
    StopWord,
    /// the token is a separator,
    /// meaning that it shouldn't be indexed but used to determine word proximity
    Separator(SeparatorKind),

>     Any,

Reviewer 4: Is it a well chosen word? Wouldn't `Other` or `Unknown` be better names?

Author: it's true that `Other` would be better because `Any` suggests that the Token could be a `Word` or a `Separator`. `Other` suggests that it's none of the previous cases.
But, in our implementation of the tokenizer, we define a `Word` as a Token which is not a `Separator` or a `StopWord`. The `Any` case is here because the creation of the token and his classification is not made at the same time, and between these two steps we set the `token_kind` to `Any`. The final user of the tokenizer should never have a token set to `Any` unless the classification is skipped.

Author: ```suggestion
    Unknown,
```
@MarinPostma, what do you think about using `Unknown` instead of `Any`? Which is clearer.

}
```

> The emitted token position are relative to the original text, this information is reliable. The `word` is normalized is need be.

Reviewer 4: ```suggestion
The emitted token position are relative to the original text, this information is reliable. The `word` is normalized if need be.
```

    // None if the token is not a separator
    pub fn is_separator(&self) -> Option<SeparatorKind> { unimplemented!() }

    // return true if the TokenKind of the token is TokenKind::StopWord
    pub fn is_stopword(&self) -> bool { unimplemented!() }
}
```

#### Internal Tokenizer trait

The `InternalTokenizer` traits provides a common interface to adapt other tokenizers to the tokenizer. This allows extensibility of the current tokenizer to other languages.

```rust
/// iterator over tokens processed by the specialized tokenizer
pub struct TokenStream<'a> {
    pub(crate) inner: Box<dyn Iterator<Item = Token<'a>> + 'a>
}

impl<'a> Iterator for TokenStream<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next()
    }
}

/// trait defining an internal tokenizer,
/// an internal tokenizer should be a script specialized tokenizer,
/// this should be implemented as an `Iterator` with `Token` as `Item`,
pub trait Tokenizer: Sync + Send {
    /// create the tokenizer based on the given `text` and `char_index`
    fn tokenize<'a>(&self, s: &'a ProcessedText<'a>) -> TokenStream<'a>;
}
```

> 
> 
> 

Reviewer 4: ```suggestion

```

- use of jieba for chinese tokenization
- use of unicode-segmenter for other segmentations

### Corner Cases

In some languages like Chinese, there can be multiple "words" extracted that are considered to be at the same position in the text. For example 计算所 gives 计算 and 计算所, that are at the same position in the text but don't have the same length, the tokenizer should support that behavior.
> return several token as the same word_position seems to be over-engineering, we may want to have only 1 token by iteration making API simplier
## Future possibilities

- We should add a way to configure tokenizer forcing a specific language/script

> - We should add a way to configure tokenizer whitelisting/backlisting separators

Reviewer 4: ```suggestion
- We should add a way to configure tokenizer whitelisting/blacklisting separators
```

  > Japanese and Korean

**lang/script detection:**
- > **[whatlang](https://github.com/greyblake/whatlang-rs):**
  > whatlang is able to detect script or/and language in a text,
  > language detection is low but the script is acceptable.
  > note: Sonic also uses whatlang to perform the tokenization, it could be interesting to check out how they do it.
- > **toku (@qdequele):**
  > in a R&D project, @qdequele was able to detect language based on stop word distribution in a text.
  > If, in a latin script, there is lot of French stop words then the text language is probably french.

**other solution that advertise multilingual support:**
- Sonic uses whatlang to perform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer
	- Sonic uses whatlang to detect the languages but doesn't actually seem to use it to segment the text. It simply uses unicode segmentation, I can't really explain what they actually do with the language information.
- tantivy advertise good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer

> 	- Tantivy is similar to elastic in the sense that you can setup a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a collection of tokenizer to choose from. Token are rather simple and do not contain any metadata, except for it's position.

Reviewer 5: ```suggestion
	- Tantivy is similar to elastic in the sense that you can set up a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a collection of tokenizer to choose from. Tokens are rather simple and do not contain any metadata, except for their position.
```

### Explanation

In order to use the tokenizer, all the user has to do is to instantiate a `Tokenizer`, call `tokenize(&str)` on it and iterate over the emitted tokens:

```rust
let stop_words = Set::from_iter(["the", "of"].iter()).unwrap();
let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));

let orig = "The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!";
let analyzed = analyzer.analyze(orig);
let mut analyzed = analyzed.tokens();
assert_eq!("the", analyzed.next().unwrap().text());
assert_eq!(" ", analyzed.next().unwrap().text());
assert_eq!("quick", analyzed.next().unwrap().text());
assert_eq!(" ", analyzed.next().unwrap().text());
assert_eq!("(", analyzed.next().unwrap().text());
assert_eq!("\"", analyzed.next().unwrap().text());
assert_eq!("brown", analyzed.next().unwrap().text());
```

> The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep it's configuration state and allocations.

Reviewer 5: ```suggestion
The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep its configuration state and allocations.
```

// original
fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
    for (_key, value) in record.iter_mut() {
        let old_value = mem::take(value);
        for (token_type, token) in simple_tokenizer(&old_value) {
            if token_type == TokenType::Word {
                let lowercase_token = token.to_lowercase();
                let to_highlight = words.contains(&lowercase_token);
                if to_highlight { value.push_str("<mark>") }
                value.push_str(token);
                if to_highlight { value.push_str("</mark>") }
            } else {
                value.push_str(token);
            }
        }
    }
}
```

As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make its API close to the previous one.

### Impact on documentation

This feature should not impact meilisearch users' documentation. 
In future versions, we will probably provide a way to configure tokenizer and this will be discussed in a new specification.

## Technical Specifications

### Architecture

The new version of the tokenizer will replace the current version as a standalone library.

> 

Author: ```suggestion

![Tokenizer](https://user-images.githubusercontent.com/6482087/102896344-8560d200-4466-11eb-8cfe-b4ae8741093b.jpg)
```

- > **[Jieba](https://github.com/messense/jieba-rs):**
  > tokenizer specialized in Chinese languages
- > **[Lindera](https://github.com/lindera-morphology/lindera):**
  > Japanese and Korean

**lang/script detection:**
- > **[whatlang](https://github.com/greyblake/whatlang-rs):**
  > whatlang is able to detect script or/and language in a text,
  > language detection is low but script is acceptable.
  > note: Sonic also uses whatlang to peform the tokenization, it could be interesting to checkout how they do it.
- > **toku (@qdequele):**
  > in a R&D project, @qdequele was able to detect language based on stop word distribution in a text.
  > If, in a latin script, there is lot of French stop words then the text language is probably french.

**other solution that advertise multilingual support:**
- Sonic uses whatlang to peform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer
	- Sonic uses whatlang to detect the languages, but don't acutally seems to use it to segment the text. It simply uses unicode segmentation, I can't really explain what they actually do with the language information.

> - tantivy advertise good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer

Reviewer 5: ```suggestion
- Tantivy advertises good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer
```

	- Tantivy is similar to elastic in the sense that you can setup a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a colleciton of tokenizer to choose from. Token are rather simple and do not contain any metadata, except for it's position.

> - How elastic search handle it: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html:

Reviewer 5: ```suggestion
- How elastic search handles it: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html:
```

> 	- Elastic proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the following components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)

Reviewer 5: "Elastic proposed" ->> is that something they don't provide anymore.

Reviewer 5: ```suggestion
	- Elastic proposed to create a custom text analyzer. A text analyzer is a pipelined text processor with the following components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizers that can be chosen, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to choose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)
```

> 	- Elastic proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the following components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)

Reviewer 5: "I also think that advanced users should be able to choose the tokenizer they want." -> Should we use "I" in the spec? I will not be able to see the author of the spec once it's merged, right?
Maybe precise who is "I" here.

Reviewer: this is a draft for now, it's still discussions we are having between us. It will eventually converge to something cleaner. I will put this is draft.

Reviewer 5: Ok, thanks! I understand this is not the final paper now. Sorry for the useless correction then!

Reviewer: no worries :)

- Algolia:

>     - Algolia use multiple techniques to handle tokenization. They start by [normalize](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the data (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurials. 

Reviewer 5: ```suggestion
    - Algolia uses multiple techniques to handle tokenization. They start by [normalize](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the data (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automatically detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurals. 
```

Reviewer 5: Oh, github does not display my changes...

### Explanation

In order to use the tokenizer, all the user has to do is to instantiate a `Tokenizer`, call `tokenize(&str)` on it and iterate over the emitted tokens:

```rust
let config = TokenizerConfig::default().set_stopwords(&["the", "of"]);
let tokenizer = Tokenizer::new(config);
let groups = tokenizer.tokenize("The quick brown fox jumps over the lazy dog");
let token_group = groups.next();
assert_eq!(token_group.tokens().map(|t| t.text()).collect(), ["The"]);
assert_eq!(token_group.normalized().map(|t| t.text()).collect(), ["the"]);
let token_group = tokens.next();
assert_eq!(token_group.tokens().map(|t| t.text()).collect(), [" "])
let token_group = groups.next();
assert_eq!(token_group.tokens().map(|t| t.text()).collect(), ["quick"])
```

> The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep it's configuration state and allocations.

Reviewer 5: ```suggestion
The call to the tokenize method allows the reuse of the same `Tokenizer` instance, and keep its configuration state and allocations.
```

## Future possibilities

- The field `char_index` is usefull for the current meilisearch but not in the future core-engine, It would be deprecated in future versions
- We should add a way to configure tokenizer forcing a specific language/script
- We should add a way to configure tokenizer adding custom stopword
- We should add a way to configure tokenizer activating/deactivating stopword detection
- We should add a way to configure tokenizer whitelisting/backlisting separators
- The tokenizer specified here is based on scripts, we should base it on languages to be able to have default stop-words for each language
- The chinese tokenizer is a complicated subject. The first implementation will simply adapt jieba's `cut` method. In another [specification](https://github.com/meilisearch/specifications/pull/5), we'll think about improving this, and this will probably require the help of a native mandarin speaker input.

> - We will want in the future to allow user configuration for the tokenizer. This is taken into account in the design of the new Tokenizer.

Reviewer 3: ```suggestion
- We will want in the future to allow user configuration for the tokenizer. This is taken into account in the design of the new Tokenizer.
- Add a normalization step before the tokenization (lowercase, remove all diacritics, remove punctuation within words, transform traditional Chinese to modern). 
```

        for (token_type, token) in simple_tokenizer(&old_value) {
            if token_type == TokenType::Word {
                let lowercase_token = token.to_lowercase();
                let to_highlight = words.contains(&lowercase_token);
                if to_highlight { value.push_str("<mark>") }
                value.push_str(token);
                if to_highlight { value.push_str("</mark>") }
            } else {
                value.push_str(token);
            }
        }
    }
}
```

#### Store

```rust
let tokenizer = Tokenizer::new(&content);
for (pos, token_group) in tokenizer.filter_map(only_token).enumerate().take(MAX_POSITION) {
    let position = (attr as usize * MAX_POSITION + pos) as u32;
    for token in token_group.normalized() {
        words_positions.entry(token.txt()).or_insert_with(SmallVec32::new).push(position);
    }
}
```

```rust
// original
for (pos, token) in simple_tokenizer(&content).filter_map(only_token).enumerate().take(MAX_POSITION) {
    let word = token.to_lowercase();
    let position = (attr as usize * MAX_POSITION + pos) as u32;
    words_positions.entry(word).or_insert_with(SmallVec32::new).push(position);

> }

Reviewer: that should be in the second section

Author: I don't think so because it an usage of the tokenizer and not his implementation

pub type Script = whatlang::Script;

/// atomic item returned by `Tokenizer::next()`
pub struct TokenGroup<'a> {
    tokens: Cow<'a, '[Token<'a>]>,
    script: Script,
}

struct NormalizedTokenIter {
    script: Script,
    token: Iter<Token<'a>>
}

> impl Deref for NormalizedToken {
> 
> }

Reviewer: ```suggestion
impl<'a> Deref for NormalizedToken<'a> {
type Target = Token<'a>;
}
```
